elastalert/test_rule.py

```diff
@@ -23,63 +36,244 @@ def print_terms(terms, parent):
+    except Exception as e:
+        print("Error querying Elasticsearch:")
+        print(repr(e)[:2048])
+        return None
+
+    num_hits = res['count']
+    print("Got %s hits from the last %s day%s" % (num_hits, args.days, 's' if args.days > 1 else ''))
+    print_terms(terms, '')
+
+    # Check for missing keys
+    pk = conf.get('primary_key')
+    ck = conf.get('compare_key')
+    if pk and not lookup_es_key(terms, pk):
+        print("Warning: primary key %s is either missing or null!")
+    if ck and not lookup_es_key(terms, ck):
+        print("Warning: compare key %s is either missing or null!")
+
+    include = conf.get('include')
+    if include:
+        for term in include:
+            if not lookup_es_key(terms, term) and '*' not in term:
+                print("Included term %s may be missing or null" % (term))
+
+    for term in conf.get('top_count_keys', []):
+        if term not in terms:
+            print("top_count_key %s may be missing" % (term))
+    print('')
+
+    # Download up to 10,000 documents to save
+    if args.save and not args.count:
+        try:
+            res = es_client.search(index, size=10000, body=query, ignore_unavailable=True)
+        except Exception as e:
+            print("Error running your filter:")
+            print(repr(e)[:2048])
+            return None
+        num_hits = len(res['hits']['hits'])
+        print("Downloaded %s documents to save" % (num_hits))
+        return res['hits']['hits']
+
+    return None
+
+
+def mock_count(rule, start, end, index):
+    """ Mocks the effects of get_hits_count using global data instead of Elasticsearch """
+    count = 0
+    for doc in data:
+        if start <= ts_to_dt(doc[rule['timestamp_field']]) < end:
+            count += 1
+    return {end: count}
+
+
+def mock_hits(rule, start, end, index):
+    """ Mocks the effects of get_hits using global data instead of Elasticsearch. """
+    docs = []
+    for doc in data:
+        if start <= ts_to_dt(doc[rule['timestamp_field']]) < end:
+            docs.append(doc)
 
-    for filename in args.files:
-        with open(filename) as fh:
-            conf = yaml.load(fh)
-        load_options(conf)
-        print("Successfully loaded %s\n" % (conf['name']))
+    # Remove all fields which don't match 'include'
+    for doc in docs:
+        for field in doc.keys():
+            if field != '_id':
+                if not any([re.match(incl.replace('*', '.*'), field) for incl in rule['include']]):
+                    doc.pop(field)
 
-        if args.schema_only:
+    # Separate _source and _id, convert timestamps
+    resp = [{'_source': doc, '_id': doc['_id']} for doc in docs]
+    [doc['_source'].pop('_id') for doc in resp]
+    replace_hits_ts(resp, rule)
+    return resp
+
+
+def mock_terms(rule, start, end, index, key, qk=None, size=None):
+    """ Mocks the effects of get_hits_terms using global data instead of Elasticsearch. """
+    if key.endswith('.raw'):
+        key = key[:-4]
```

Comment:
`key.rsplit('.raw', 1)[0]` will also do the trick. But given the explicit use of `.endswith` here, chopping off the last four characters is a safe operation.